### ISSUE TITLE
weird behavior for `include_initial` changes on a stream ordered by a secondary index

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#5153]
Described by tatsujin in the Slack channel.  His query is `r.db('thedatabase').table('thetable').order_by(index=r.asc('_modified')).filter(lambda var_1: r.expr(['abc', 'def', 'ghi']).contains(var_1['aaa'])).filter(lambda var_2: (~r.expr(['P', 'R']).contains(var_2['bbb'].default([])))).filter(lambda var_3: r.row.has_fields('xxx', 'yyy')).filter(lambda var_4: (r.row['time'] > (r.now() - r.expr(7200)))).pluck(['time', 'xxx', 'yyy', '_key', 'tags', 'clear', 'bbb', 'state', '_key'])` and he's occasionally receiving initial and uninitial vals of `None`.  Those vals go away if he removes the `order_by`.

This is strange (although it's possible I'm forgetting something obvious).

**Comment by tatsujin1** [comment]
Should add that the `None` events only occurrs when changes are simultaneously being made to the table (by another process).

**Comment by tatsujin1** [comment]
I've made a self-contained test that reproduces the problem:
*test-modify.py*
```
#! /usr/bin/env python3

import rethinkdb as r
import random
import time

r.connect().repl()

db_name = 'test_db'
table_name = 'test_table'

if db_name not in r.db_list().run():
	r.db_create(db_name).run()

if table_name not in r.db(db_name).table_list().run():
	r.db(db_name).table_create(table_name).run()
	
table = r.db(db_name).table(table_name)
table.delete().run()
	
indexes = ['_modified']
existing_indexes = table.index_list().run()
for idx in indexes:
	if idx not in existing_indexes:
		table.index_create(idx).run()

max_id = 5000
print('filling table...', max_id)
for doc_id in range(1, max_id + 1):
	table.insert({
		'id': doc_id,
		'_modified': r.now(),
		'value': random.randint(1, 10)
	}).run()



print('making changes...')
while True:
	time.sleep(0.05)

	doc_id = random.randint(1, max_id)
	doc = {
		'_modified': r.now(),
		'value': random.randint(1, 10)
	}

	existing = table.get(doc_id).run()
	if existing:
		if random.randint(1, 100) < 10:
			table.get(doc_id).delete().run()
			print('   del:', doc_id)
		else:
			table.get(doc_id).update(doc).run()
			print('  upd :', doc_id)

	else:
		doc['id'] = doc_id
		table.insert(doc).run()
		print('ins   :', doc_id)
```
Start that and leave it running.
Then run this, *test-listen.py*:
```
#! /usr/bin/env python3

import rethinkdb as r

r.connect().repl()

db_name = 'test_db'
table_name = 'test_table'

table = r.db(db_name).table(table_name)
	
print('receiving changes...')
missing = object()
for change in table.order_by(index=r.asc('_modified')).changes(include_initial=True, include_states=True).run():
	state = change.get('state')
	if state:
		if state == 'ready':
			print('initial batch end')
			break
		continue

	new = change.get('new_val', missing)
	old = change.get('old_val', missing)

	if len(change) == 1 and (old is None or new is None):
		print('got None event: %s' % repr(change))
		continue

	if new is missing:
		print('got "uninitial": %s' % repr(change))
		
	elif old is missing:
		#print('initial:', new['id'])
		pass

	elif old is None:
		print('create :', new['id'])
							
	elif new is None:
		print('delete :', old['id'])
	
	else:
		print('change :', new['id'])
```
It will receive events until `state='ready'`. It will print 'got None event' when the "trouble" arises.
It doesn't happen always, so multiple runs might be required.

**Comment by tatsujin1** [comment]
Forgot to mention that it seems that only "create" causes those `None` events to be emitted.
If the above "test-modifier.py" is changed to do only `insert` it never happens.

**Comment by mlucy** [comment]
Thanks @tatsujin1 !  That's a huge help.

**Comment by tatsujin1** [comment]
Has anyone tried reproducing this?

**Comment by danielmewes** [comment]
@tatsujin1 Just tried and I can reproduce it easily with your scripts! Thanks again for providing those. That will make debugging this a lot easier.

**Comment by tatsujin1** [comment]
Nice :)

**Comment by tatsujin1** [comment]
huh? how/where did I reference that issue?

**Comment by mlucy** [comment]
A fix for this is in CR 3400 by @danielmewes.

@tatsujin1 -- it looks like it is in fact safe to just totally ignore these changes.  They're caused by deletes to parts of the range that you haven't read initial values for yet: the changefeed logic responsible for discarding changes that don't apply to the current active range was just being slightly too liberal and passing them through (but it was conservative to strip out the `old_val`, since it was outside the active range, resulting in the strange `{new_val: null}` documents).

**Comment by tatsujin1** [comment]
Oh, I see, partly :). Does this explanation/fix also include the other variant: `{ old_val: None }` ?

**Comment by mlucy** [comment]
Yup (`{old_val: None}` would be caused by inserts to parts of the range you haven't read initial values for yet).

**Comment by tatsujin1** [comment]
Ok, just checking :)

**Comment by mlucy** [comment]
This is in next and 2.2.x.

**Comment by danielmewes** [comment]
@tatsujin1 We've released RethinkDB 2.2.2 today which contains the fix for this. Would you mind upgrading and letting us know if you're still seeing any issues?

**Comment by tatsujin1** [comment]
Confirmed, this doesn't happen any more!
